src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,147 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(last, value, null);
+        if (isEmpty()) {
+            first = newNode;
+        } else {
+            last.next = newNode;
+        }
+        last = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, first);
+            newNode.next = first;
+            first.prev = newNode;
+            first = newNode;
+        } else {
+            Node<T> node = getNodeIndex(index);
+            Node<T> newNode = new Node<>(node.prev, value, node);
+            node.prev.next = newNode;
+            node.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T e : list) {
+            add(e);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        return getNodeIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> oldNode = getNodeIndex(index);
+        T item = oldNode.item;
+        oldNode.item = value;
+        return item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        Node<T> nodeToRemove = getNodeIndex(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = first;
+        while (node != null) {
+            if (isEquals(object, node)) {
+                unlink(node);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
+
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        if (first == null) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private boolean isEquals(T object, Node<T> node) {
+        return object == node.item || object != null && object.equals(node.item);
+    }
+
+    private Node<T> getNodeIndex(int index) {
+        isValidIndex(index);
+        Node<T> current = first;
```

Comment:
Implement according this suggestion from the common mistakes please:
"Don’t iterate your nodes from the head to tail or vice versa.
Pay attention to the index you receive as an input, you can iterate only first or second half of the list, depending on the index value. That will boost your lists’ performance."

Reply:
Not fixed, please pay attention to the @maxline comment